signac/contrib/project.py

```diff
@@ -135,36 +115,29 @@ class Project:
     def __init__(self, config=None, _ignore_schema_version=False):
         if config is None:
             config = load_config()
-        self._config = _ProjectConfig(
-            config, _mutate_hook=partial(_invalidate_config_cache, self)
-        )
+        self._config = _ProjectConfig(config)
         self._lock = RLock()
 
-        # Prepare cached properties derived from the project config.
-        self._id = None
-        self._rd = None
-        self._wd = None
-
         # Ensure that the project id is configured.
-        if self.id is None:
+        try:
+            self._id = str(self.config["project"])
+        except KeyError:
             raise LookupError(
                 "Unable to determine project id. "
                 "Please verify that '{}' is a signac project path.".format(
                     os.path.abspath(self.config.get("project_dir", os.getcwd()))
                 )
             )
 
-        # Ensure that the project's data schema is supported.
-        if not _ignore_schema_version:
-            self._check_schema_compatibility()
-
-        # Prepare project document
-        self._document = None
-
-        # Prepare project H5StoreManager
-        self._stores = None
+        # Prepare root directory and workspace paths.
+        self._root_directory = self.config["project_dir"]
```

Comment:
Do we already have an issue to start using `pathlib.Path` for these objects?

Reply:
I did some benchmarking at one point. Switching to `Path` objects affects performance somewhat significantly. The construction and manipulation of `Path` objects involve a lot of system calls that do not happen if using the `os` library functions like `os.path.join`, etc. I would advocate keeping the current "string path" approach.

Reply:
See also: https://github.com/glotzerlab/signac/pull/511#issuecomment-782292032

Reply:
Ok, good to know. 👌

Reply:
@bdice Considering that this question might come up again. Would you mind adding a comment into the code so that future developers are aware?

Reply:
@csadorf Where should the comment go?

Reply:
Since a future change would likely need to inspect these variables:
https://github.com/glotzerlab/signac/blob/a5f6c3048c1cd9e1defc1c5521ba2e49bf2e3e43/signac/contrib/project.py#L266-L267
Either right there or above the `import os` statement:
https://github.com/glotzerlab/signac/blob/a5f6c3048c1cd9e1defc1c5521ba2e49bf2e3e43/signac/contrib/project.py#L10
We could also add an explicit `import os.path` import statement and add the comment right there.

Reply:
It is generally a good idea to be pretty explicit in the code about what are critical paths, e.g., like here:
https://github.com/glotzerlab/signac/blob/a5f6c3048c1cd9e1defc1c5521ba2e49bf2e3e43/signac/contrib/project.py#L2587-L2588

Reply:
@csadorf Thanks for the suggestion, you're absolutely right that this is important. I improved this in 0e1be15c8ae0fae660eb341ca4accddbfaf68859 and gave greater explanation for a few optimizations that are not obvious.